File: 2024/03 March/Print N-bit binary numbers having more 1s than 0s/main.cpp
// Link: https://www.geeksforgeeks.org/problems/print-n-bit-binary-numbers-having-more-1s-than-0s0252/1

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    void solve(int n, int n0, int n1, int tot, string s, priority_queue<pair<int, string>> &pq)
    {
        if (n == 0)
        {
            pq.push({tot, s});
            return;
        }

        solve(n - 1, n0, n1 + 1, tot * 2 + 1, s + "1", pq);
        if (n1 > n0)
        {
            solve(n - 1, n0 + 1, n1, tot * 2, s + "0", pq);
        }
    }

    vector<string> NBitBinary(int n)
    {
        priority_queue<pair<int, string>> pq;
        solve(n, 0, 0, 0, "", pq);

        vector<string> v;

        while (!pq.empty())
        {
            v.push_back(pq.top().second);
            pq.pop();
        }

        return v;
    }
};

int main()
{
    Solution obj;
    int n = 2;
    vector<string> ans = obj.NBitBinary(n);
    for (auto x : ans)
    {
        cout << x << " ";
    }
    cout << endl;
    n = 3;
    ans = obj.NBitBinary(n);
    for (auto x : ans)
    {
        cout << x << " ";
    }
    cout << endl;
    return 0;
}

File: 2024/03 March/Print N-bit binary numbers having more 1s than 0s/main.py
# Link: https://www.geeksforgeeks.org/problems/print-n-bit-binary-numbers-having-more-1s-than-0s0252/1


class Solution:
    def NBitBinary(self, n):
        ans = []
        self.backtrack(["1"], 1, n, ans)
        return ans

    def backtrack(self, arr, i, n, ans):
        if len(arr) == n:
            ans.append("".join(arr))
        else:
            arr.append("1")
            self.backtrack(arr, i + 1, n, ans)
            if i > 0:
                arr[-1] = "0"
                self.backtrack(arr, i - 1, n, ans)
            arr.pop()


if __name__ == "__main__":
    obj = Solution()
    n = 2
    print(obj.NBitBinary(n))
    n = 3
    print(obj.NBitBinary(n))
